### DISCUSSION TITLE
Gradcam 3D Visualisation

### DISCUSSION THREAD

**Discussion by Meddebma** [Project-MONAI/MONAI#2360] [Q&A]
Hello, 

I was trying to implement occlusion sensitivity map and Gradcam for my classification task but I've got two Problems:
1- the displayed slice was in a different reconstruction as the original one(displayed in Check Data Shape section), how can I make the same reconstruction?

2- for the Gradcam I got this error: 
AttributeError: 'numpy.ndarray' object has no attribute 'detach'

here is the code:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification.ipynb

Thank you very much

**Comment by Nic-Ma** [comment]
Hi @Meddebma ,

Thanks for your interest here.
Let me try to reproduce your issue locally and get back to you as soon as possible.

Thanks.

**Reply by rijobro** [reply]
1. In your section `Check Data Shape`, you use `Orientation(RPI)`. You don't use this elsewhere in your code, so all other visualisations of your images may be in a different orientation.

2. `GradCAM` returns the result as a numpy image, so `.detach()` doesn't exist. This is why in [MONAI's GradCAM tutorials](https://github.com/Project-MONAI/tutorials/blob/master/modules/interpretability/class_lung_lesion.ipynb) we use:

```python
if isinstance(im, torch.Tensor):
    im = im.cpu().detach()
```

**Reply by Nic-Ma** [reply]
Hi @rijobro ,

I checked the issue again, we converted the Tensor to numpy at:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/visualize/class_activation_maps.py#L34
I would make a PR to enhance this `default_normalizer` to use the same output data type as input data.
What do you think?

Thanks.

**Reply by Nic-Ma** [reply]
Hi @rijobro and @Meddebma ,

I submitted PR to enhance the visualization part.
https://github.com/Project-MONAI/MONAI/pull/2368

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma, 
thank you very much! how can I use the PR in my code now?

**Reply by Nic-Ma** [reply]
The PR had been merged, you can use the dev branch of MONAI directly.

Thanks.

**Reply by Meddebma** [reply]
@Nic-Ma 
Perfect! now I can vizualise it! The CT-Slice is curiously still in another reconstruction than the original one (RAS), how can I display the CT-Slice as it is originally?
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification.ipynb

**Reply by Nic-Ma** [reply]
You can consider to use dict data as input and use `Invertd` transform to invert all the spatial transforms:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/post/dictionary.py#L417

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma 
I tried the invertd but It changed nothing, so I tried Orientation transform, and I've got this error:
AttributeError                            Traceback (most recent call last)

<ipython-input-35-e9a9acb978e1> in <module>()
      1 # Get a random image and its corresponding label
----> 2 img, label = get_next_im()
      3 
      4 # Get the occlusion sensitivity map
      5 occ_sens = monai.visualize.OcclusionSensitivity(

<ipython-input-34-0cf50e03166d> in get_next_im()
      9 def get_next_im():
     10     test_data = next(itera)
---> 11     return test_data[0].to(device), test_data[1].unsqueeze(0).to(device)
     12 
     13 

AttributeError: 'list' object has no attribute 'to'

https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification.ipynb
thanks.

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

As I said in the previous thread, please try to use **dict data as input** and **dict transforms**, then use `Invertd` transform to invert.
You can refer to this example:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_inference_dict.py

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma , 
I implemented it like this, but somehow I failed to trasform the data into dict data
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification.ipynb

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

1. Make your input data as a dict.
2. Use transforms with "d" as pre_transforms, like "Resized()", "Spacingd()", etc.
3. Use "Dataset" instead of "ImageDataset", and use above pre_transforms in the dataset.
4. Post transforms should also use the transforms with "d", use "InvertD" in the post transforms.

Please refer to the example for more details:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_inference_dict.py

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma, 

I followed the steps, I think I still have an error in the implementation of post transforms, can you please take a look:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification_Dict.ipynb

thank you.

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

I don't quite understand your code:
```py
val_ds = monai.data.Dataset(data=val_files, transform=val_transforms)
val_loader = DataLoader(val_ds, batch_size=2, num_workers=4, pin_memory=torch.cuda.is_available())

with torch.no_grad():
  for d in val_loader:
         post_transforms(d)
         itera = iter(d)
```
Seems you didn't run model prediction at all, and you didn't save the pred in a dictionary with key "pred".

Thanks.

**Reply by Meddebma** [reply]
@Nic-Ma 
In the example you sent me was a segmentation task, for my classification task I thought I dont need to make a prediction and save it, or do you think I should do the same?
Thanks

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

Sorry I misunderstood your problem last week.
Seems you want to invert the input image, right? If so, please change the expected `keys` of `Invertd` transform to image instead of the `pred` which is not existing in your program at all.

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma,

so I did change it to image but I've got this Error:

Return test_data[0].to(device), test_data[1].unsqueeze(0).to(device)
AttributeError: 'str' object has no attribute 'to'

https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification_Dict.ipynb

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

Sorry I didn't get what you suppose to do in this code:
```py
with torch.no_grad():
  for d in val_loader:
         post_transforms(d)
         itera = iter(d)

def get_next_im():
    test_data = next(itera)
    return test_data[0].to(device), test_data[1].unsqueeze(0).to(device)
```
@wyli @ericspod @rijobro do you guys have any ideas? Maybe I misunderstood the question?

Thanks in advance.

**Reply by Meddebma** [reply]
@Nic-Ma 

I was trying to adapt the densenet_training array to dict, here was the code I was trying to change
https://github.com/Project-MONAI/tutorials/blob/master/3d_classification/densenet_training_array.ipynb

**Reply by rijobro** [reply]
Hi @Meddebma, I think you're better off skipping the `Invertd` initially. (I think you only need to invert if you want to save the output of your GradCAM or occlusion sensitivity to file.)

1. train with `train_loader`, which will apply `train_transforms`. You're already doing this 👍 
2. get a single image from `val_ds`, which will apply `val_transforms` e.g., `x = val_ds[0]["image"]`.
3. add a batch dimension and move to device e.g., `x = x.unsqueeze(0).to(device)`.
4. send to GradCAM and/or occlusion sensitivity.
5. visualise the original image (`x`) and any of your GradCAM/occlusion sensitivity results.

If you want to save these results to file and have them have the same size/orientation as your input image, then you would apply the `Invertd` at this point, but first you want to get points 1-5 working properly.

**Reply by Meddebma** [reply]
Hi @rijobro, 
thank you! since I also need the label for occlusion sensitivity, I tried this
y = val_ds[1]["label"]
y = y.unsqueeze(0).to(device)
but it did not work out, I've got this error: AttributeError: 'str' object has no attribute 'to'

**Reply by rijobro** [reply]
are you sure you used `val_ds` and not `val_files`? `val_ds` should give you the image with `val_transforms` applied to it, whereas `val_files` will give you the filename. This would be as a string, and would explain the error you're seeing.

**Reply by rijobro** [reply]
Sorry ignore my previous comment, I didn't realise you were talking about `label`. Label doesn't need to be unsqueezed and doesn't need to be moved to the device. It should just be a single number (0 or 1 depending on the ground-truth classification class).

e.g., 
```python
data = val_ds[0]
x, y = data["image"], data["label"]
x = x.unsqueeze(0).to(device)
print("this is the label: ", y)
```

**Reply by Meddebma** [reply]
@rijobro 
thank you, It worked this time! 
Somehow I feel that the results of cam and occlusion sensitivity are divergent, could it be the case? And is it possible to transpose the original image with the gradcam? 
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification_Dict.ipynb

**Reply by rijobro** [reply]
GradCAM and occlusion sensitivity are different, so aren't guaranteed to always highlight the same differences. In your case, you only used one example, so it's hard to figure out any trends based on that. Try adding more rows to your visualisation (and use random numbers rather than `val_ds[2]`). 

Your confusion matrix also highlights that your network isn't currently doing very well, so you probably want to improve that before the gradcam and occlusion sensitivity become useful. Just to explain that a little further, there's no point asking your network "why did you make that decision?" if it's currently not making very educated guesses.

**Reply by Meddebma** [reply]
@rijobro 
Thank you I did that and it worked well! 
I am not sure that the correct labels belong to the correct patients, how can I best find out if it fits? can I display the name of the nifti files with the labels? the labels are in a csv with the names of the nifti files

**Reply by rijobro** [reply]
I looked through your code and it looks ok. This would be ok:

```python
for data in data_loader:
    x, y = data["img"], data["label"]
```
because you get both the label and the image at the same time. 

If you have an error it would therefore be at the reading of your excel sheet versus the glob for your files. This starts before the MONAI workflow, so you'll have to check this yourself. For example, when you run `sorted(glob(os.path(...))` are the files returned in the order you that you expect?

**Reply by Meddebma** [reply]
@rijobro Okey I double checked and it looks fine with sorted(glob...) :)
By the way in the confusion matrix, I wanted to implement other metrics using this:
cmm = ConfusionMatrixMetric (include_background=True, metric_name=["sensitivity","specificity","negative predictive value"])
how can I display the results?
